binrw_derive/src/codegen/typed_builder.rs

```diff
@@ -0,0 +1,248 @@
+use proc_macro2::TokenStream;
+use quote::{quote, ToTokens};
+use syn::{Ident, Type};
+
+#[allow(clippy::wildcard_imports)]
+use crate::codegen::sanitization::*;
+use crate::parser::meta_types::NamedImport;
+
+pub(crate) enum BuilderFieldKind {
+    Required,
+    Optional { default: Box<syn::Expr> },
+}
+
+pub(crate) struct BuilderField {
+    pub(crate) name: Ident,
+    pub(crate) ty: Type,
+    pub(crate) kind: BuilderFieldKind,
+}
+
+pub(crate) struct Builder<'a> {
+    pub(crate) builder_name: &'a Ident,
+    pub(crate) result_name: &'a Ident,
+    pub(crate) fields: &'a [BuilderField],
+}
+
+impl<'a> Builder<'a> {
+    pub(crate) fn generate(&self) -> TokenStream {
+        let builder_name = self.builder_name;
+        let name = self.result_name;
+        let fields = self.generate_result_fields();
+        let builder_fields = self.generate_builder_fields();
+        let initial = self.generate_builder_initial();
+        let generics = self.generate_generics();
+        let initial_generics = self.generate_initial_generics();
+        let setters = self.generate_setters();
+        let satisfied = std::iter::repeat(SATISFIED_OR_OPTIONAL);
+        let field_names: Vec<_> = self.fields.iter().map(|field| &field.name).collect();
```

Comment:
This one is actually needed as field_names is used multiple times and it would be consumed by the first usage. it's also possible to clone the iterator for each usage but.... I don't think that improves anything.

Reply:
cloning the iterator i think may be better, since it's only carrying a single reference to the fields instead of needing to create storage for a name reference per field.

Reply:
I mean, for efficiency it _might_ be better, but for readability I think it's too much of a loss to be worth the micro-optimization.